changes to support ATen code generation inside fbcode

anderspapitto opened the pull request.

soumith commented on torch/_thnn/utils.py:
```diff
@@ -2,8 +2,11 @@
 
-THNN_H_PATH = os.path.join(os.path.dirname(__file__), '..', 'lib', 'THNN.h')
-THCUNN_H_PATH = os.path.join(os.path.dirname(__file__), '..', 'lib', 'THCUNN.h')
+try:
```
this is the part that looks sketchy, worth a comment. otherwise PR looks good to go.

soumith reviewed (approved):
approved with minor comment

Commit:
Back out "Back out "Add support for generating ATen files during fbcode build""

Original commit changeset: 7b8de22d1613

I'm re-sending this diff exactly as it was approved and
committed. Fixes to support @mode/opt will be sent separately for ease
of review.

Commit:
Enable building //caffe2:torch with @mode/opt

In @mode/opt, python runs out of a PAR, which breaks a lot of
assumptions in the code about where templates/ folders live relative
to __file__. Rather than introduce hacks with parutil, I simply turn
template_path into a parameter for all the relevant functions and
thread it through from the top level.
